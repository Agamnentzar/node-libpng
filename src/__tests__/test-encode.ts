import { encode, writePngFile, writePngFileSync } from "..";
import { readFileSync } from "fs";

const someGradient = Buffer.alloc(256 * 256 * 3);
for (let x = 0; x < 256; ++x) {
    for (let y = 0; y < 256; ++y) {
        const index = (y * 256 + x) * 3;
        someGradient[index + 0] = x;
        someGradient[index + 1] = y;
        someGradient[index + 2] = 0;
    }
}

const someOrangeRectangle = Buffer.alloc(16 * 16 * 3);
for (let index = 0; index < someOrangeRectangle.length; index += 3) {
    someOrangeRectangle[index + 0] = 255;
    someOrangeRectangle[index + 1] = 128;
    someOrangeRectangle[index + 2] = 64;
}

const someOpaqueRectangle = Buffer.alloc(16 * 16 * 4);
for (let index = 0; index < someOpaqueRectangle.length; index += 4) {
    someOpaqueRectangle[index + 0] = 64;
    someOpaqueRectangle[index + 1] = 128;
    someOpaqueRectangle[index + 2] = 64;
    someOpaqueRectangle[index + 3] = 128;
}

describe("encode", () => {
    it("encodes a png", () => {
        const encoded = encode(someGradient, {
            width: 256,
            height: 256,
            alpha: false,
        });
        expect(encoded.toString("hex")).toMatchSnapshot();
    });

    it("encodes a png with an alpha channel", () => {
        const encoded = encode(someOpaqueRectangle, {
            width: 16,
            height: 16,
            alpha: true,
        });
        expect(encoded.toString("hex")).toMatchSnapshot();
    });

    it("throws an error when trying to encode something which isn't a buffer", () => {
        const options = {
            width: 16,
            height: 16,
            alpha: false,
        };
        expect(() => encode("something" as any, options)).toThrowErrorMatchingSnapshot();
    });

    it("throws an error when trying to encode a PNG with an alpha channel with alpha specified as disabled", () => {
        const options = {
            width: 16,
            height: 16,
            alpha: false,
        };
        expect(() => encode(someOpaqueRectangle, options)).toThrowErrorMatchingSnapshot();
    });

    [
        {
            width: 16,
            alpha: false,
        },
        {
            height: 16,
            alpha: false,
        },
        {
            width: 16,
            height: 16,
            alpha: false,
        },
    ].forEach(options => {
        it(`encodes a PNG with ${JSON.stringify(options)} as options`, () => {
            expect(encode(someOrangeRectangle, options).toString("hex")).toMatchSnapshot();
        });
    });

    [
        {
            alpha: false,
        },
        {
            width: 20,
            alpha: "false",
        },
        {},
        null,
        undefined,
        "string",
        {
            width: "twenty",
            alpha: true,
        },
        {
            width: 0,
            height: 0,
            alpha: true,
        },
        {
            width: 20,
            height: 1.5,
            alpha: true,
        },
        {
            width: 1.5,
            height: 20,
            alpha: true,
        },
        {
            width: 1000,
            height: 1000,
            alpha: true,
        },
    ].forEach(options => {
        it(`throws an error with bad options ${JSON.stringify(options)}`, () => {
            expect(() => encode(someOrangeRectangle, options as any)).toThrowErrorMatchingSnapshot();
        });
    });
});

describe("writePngFileSync", () => {
    it("encodes a PNG and writes it to disk", () => {
        const options = { width: 16, alpha: false };
        const encoded = encode(someOrangeRectangle, options);
        const path = `${__dirname}/../../tmp-write-sync.png`;
        writePngFileSync(path, someOrangeRectangle, options);
        const fromDisk = readFileSync(path);
        expect(Array.from(fromDisk)).toEqual(Array.from(encoded));
    });
});

describe("writePngFile", () => {
    describe("using the Promise API", () => {
        it("encodes a PNG and writes it to disk", async () => {
            const options = { width: 16, alpha: false };
            const encoded = encode(someOrangeRectangle, options);
            const path = `${__dirname}/../../tmp-write-promise.png`;
            await writePngFile(path, someOrangeRectangle, options);
            const fromDisk = readFileSync(path);
            expect(Array.from(fromDisk)).toEqual(Array.from(encoded));
        });

        it("rejects with an error when encoding failed", async () => {
            const options = { width: 16, alpha: false };
            const path = `${__dirname}/../../tmp-write-promise-error-encoding.png`;
            return expect(writePngFile(path, Buffer.alloc(2), options)).rejects.toMatchSnapshot();
        });

        it("rejects with an error when writing failed", async () => {
            const options = { width: 16, alpha: false };
            const path = `${__dirname}/this-file/does/not/exist.png`;
            return expect(writePngFile(path, someOrangeRectangle, options)).rejects.toMatchSnapshot();
        });
    });

    describe("using the callback API", () => {
        it("encodes a PNG and writes it to disk", done => {
            const options = { width: 16, alpha: false };
            const encoded = encode(someOrangeRectangle, options);
            const path = `${__dirname}/../../tmp-write-callback.png`;
            writePngFile(path, someOrangeRectangle, options, error => {
                expect(error).toBeNull();
                const fromDisk = readFileSync(path);
                expect(Array.from(fromDisk)).toEqual(Array.from(encoded));
                done();
            });
        });

        it("calls the callback with an error when encoding failed", done => {
            const options = { width: 16, alpha: false };
            const path = `${__dirname}/../../tmp-write-callback-error-encoding.png`;
            writePngFile(path, Buffer.alloc(2), options, error => {
                expect(error).toMatchSnapshot();
                done();
            });
        });

        it("calls the callback with an error when writing failed", done => {
            const options = { width: 16, alpha: false };
            const path = `${__dirname}/this-file/does/not/exist.png`;
            writePngFile(path, someOrangeRectangle, options, error => {
                expect(error).toMatchSnapshot();
                done();
            });
        });
    });
});
