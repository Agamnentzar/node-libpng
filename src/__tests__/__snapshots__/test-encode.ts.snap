// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode encodes a PNG with {"height":8,"alpha":false} as options 1`] = `"89504e470d0a1a0a0000000d49484452000000100000000808020000007f14e8c0000000174944415418d363fcdfe0c0400a606220118c6a20060000b79f01cf914739e30000000049454e44ae426082"`;

exports[`encode encodes a PNG with {"width":16,"alpha":false} as options 1`] = `"89504e470d0a1a0a0000000d49484452000000100000000808020000007f14e8c0000000174944415418d363fcdfe0c0400a606220118c6a20060000b79f01cf914739e30000000049454e44ae426082"`;

exports[`encode encodes a PNG with {"width":16,"height":8,"alpha":false} as options 1`] = `"89504e470d0a1a0a0000000d49484452000000100000000808020000007f14e8c0000000174944415418d363fcdfe0c0400a606220118c6a20060000b79f01cf914739e30000000049454e44ae426082"`;

exports[`encode encodes a png 1`] = `"89504e470d0a1a0a0000000d4948445200000100000001000802000000d3103f31000002024944415478daedd3010900000c024185f5cfbc026bb03b4c207c93d4eceb260dbc250004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200012000012000100008000400020001800040002000100008000400020001800040002000100008000400020001800040002000100008000400020001800040002000100008000400020001800040002000100008000400020001800040002000100008000180004000200010000800040002000180004000200010000800040002000180004000200010000800040002000180004000200010000800040002000180004000200010000800040002000180004000200010000800040002000180004000084000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000240004e400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000200018000400020001000080004000280cb021a6505fccfd059aa0000000049454e44ae426082"`;

exports[`encode encodes a png with an alpha channel 1`] = `"89504e470d0a1a0a0000000d49484452000000100000001008060000001ff3ff610000001d4944415438cb637468706860a00030315008460d183560d480c1620000596a01a0768b6c110000000049454e44ae426082"`;

exports[`encode throws an error when trying to encode a PNG with an alpha channel with alpha specified as disabled 1`] = `"Error encoding PNG.Invalid buffer length."`;

exports[`encode throws an error when trying to encode something which isn't a buffer 1`] = `"Input is not a buffer."`;

exports[`encode throws an error with bad options "string" 1`] = `"Options need to be an object."`;

exports[`encode throws an error with bad options {"alpha":false} 1`] = `"Error encoding PNG. Either width or height need to be specified."`;

exports[`encode throws an error with bad options {"width":"twenty","alpha":true} 1`] = `"Error encoding PNG. Either width or height need to be specified."`;

exports[`encode throws an error with bad options {"width":0,"height":0,"alpha":true} 1`] = `"Error encoding PNG.Invalid buffer length."`;

exports[`encode throws an error with bad options {"width":1.5,"height":20,"alpha":true} 1`] = `"Error encoding PNG. Width needs to be an integer."`;

exports[`encode throws an error with bad options {"width":20,"alpha":"false"} 1`] = `"Error encoding PNG. Alpha channel needs to be explicitly specified."`;

exports[`encode throws an error with bad options {"width":20,"height":1.5,"alpha":true} 1`] = `"Error encoding PNG. Height needs to be an integer."`;

exports[`encode throws an error with bad options {"width":1000,"height":1000,"alpha":true} 1`] = `"Error encoding PNG.Invalid buffer length."`;

exports[`encode throws an error with bad options {} 1`] = `"Error encoding PNG. Either width or height need to be specified."`;

exports[`encode throws an error with bad options null 1`] = `"Options need to be an object."`;

exports[`encode throws an error with bad options undefined 1`] = `"Options need to be an object."`;

exports[`writePngFile using the Promise API rejects with an error when encoding failed 1`] = `[TypeError: Error encoding PNG.]`;

exports[`writePngFile using the callback API calls the callback with an error when encoding failed 1`] = `[TypeError: Error encoding PNG.]`;
